vita3k/host/src/pup.cpp

```diff
@@ -0,0 +1,473 @@
+        infile.read((char *)encrypted_data, sceseg.size); //maybe rewrite the casts to static_cast<>() ??
+        aes_context aes_ctx;
+        aes_setkey_enc(&aes_ctx, (unsigned char *)sceseg.key.c_str(), 128);
+        size_t ctr_nc_off = 0;
+        unsigned char ctr_stream_block[0x10];
+        unsigned char *decrypted_data = new unsigned char[sceseg.size];
+        aes_crypt_ctr(&aes_ctx, sceseg.size, &ctr_nc_off, (unsigned char *)sceseg.iv.c_str(), ctr_stream_block, encrypted_data, decrypted_data);
+        if (sceseg.compressed) {
+            mz_stream stream;
+            stream.zalloc = Z_NULL;
+            stream.zfree = Z_NULL;
+            stream.opaque = Z_NULL;
+            stream.avail_in = 0;
+            stream.next_in = Z_NULL;
+            if (mz_inflateInit(&stream) != MZ_OK) {
+                LOG_ERROR("inflateInit failed while decompressing");
+                return;
+            }
+
+            std::string compressed_data((char *)decrypted_data, sceseg.size);
+            stream.next_in = (Bytef *)compressed_data.data();
+            stream.avail_in = compressed_data.size();
+
+            int ret;
+            char outbuffer[524288];
+            std::string decompressed_data;
+
+            do {
+                stream.next_out = reinterpret_cast<Bytef *>(outbuffer);
+                stream.avail_out = sizeof(outbuffer);
+
+                ret = mz_inflate(&stream, 0);
+
+                if (decompressed_data.size() < stream.total_out) {
+                    decompressed_data.append(outbuffer, stream.total_out - decompressed_data.size());
+                }
+
+            } while (ret == MZ_OK);
+
+            mz_inflateEnd(&stream);
+
+            if (ret != MZ_STREAM_END) {
+                LOG_ERROR("Exception during zlib decompression: ({}) {}", ret, stream.msg);
+                return;
+            }
+            outfile.write(decompressed_data.c_str(), decompressed_data.size());
+        } else {
+            outfile.write((char *)decrypted_data, sceseg.size);
+        }
+        outfile.close();
+        delete[] encrypted_data;
+        delete[] decrypted_data;
+    }
+};
+
+void join_files(std::string path, std::string filename, std::string output) {
+    std::vector<std::string> files;
+
+    for (auto &p : fs::directory_iterator(path)) {
+        if (p.path().filename().string().substr(0, 4) == filename) {
+            files.push_back(p.path().string());
+        }
+    }
+
+    std::ofstream fileout(output, std::ios::binary);
+    for (auto file : files) {
+        std::ifstream filein(file, std::ios::binary);
+        char *buffer = new char[fs::file_size(file)];
+        filein.read(buffer, fs::file_size(file));
+        fileout.write(buffer, fs::file_size(file));
+        filein.close();
+        fs::remove(file);
+    }
+    fileout.close();
+}
+
+void decrypt_pup_packages(std::string src, std::string dest) {
+    std::vector<std::string> pkgfiles;
+
+    KeyStore SCE_KEYS = KeyStore();
```

Comment:
```suggestion
    KeyStore SCE_KEYS;
```

Auto init so dont need to do that fiend,